### ISSUE TITLE
step2 failed on HPC where docker cannot be installed

### ISSUE THREAD

**Issue by SichongP** [dahak-metagenomics/taco-read-filtering#1]
Encountered following error when running step 2 of the workflow. It appears to result from docker being missing. I've been trying to work around this but couldn't find a way to get docker running without root access. Singularity is loaded and available when running.

Error msg as follows:

/bin/bash: line 2: docker: command not found
Error in rule pre_trimming_quality_assessment:
    jobid: 0
    output: data/SRR606249_1_reads_fastqc.zip

RuleException:
CalledProcessError in line 93 of /home/pengsc/dahak/taco-read-filtering/rules/read_filtering/pre_post_assessment.rule:
Command ' set -euo pipefail;
            set -x
            docker run                 -u "$(id -u):$(id -g)"                 -v /home/pengsc/dahak/taco-read-filtering/data/:/data/                 -it                 quay.io/biocontainers/fastqc:0.11.7--pl5.22.0_2                 fastqc                 -t 1                 /data/SRR606249_1_reads.fq.gz                 -o /data/
            set +x ' returned non-zero exit status 127.
  File "/home/pengsc/dahak/taco-read-filtering/rules/read_filtering/pre_post_assessment.rule", line 93, in __rule_pre_trimming_quality_assessment
  File "/home/pengsc/miniconda3/lib/python3.6/concurrent/futures/thread.py", line 56, in run

**Comment by charlesreid1** [comment]
Okay - I think I know where this is coming from. This is due to a missing `singularity:` directive in the rules files.

**Comment by charlesreid1** [comment]
I stand corrected - [the singularity directive](https://github.com/dahak-metagenomics/taco-read-filtering/blob/master/rules/read_filtering/pre_post_assessment.rule#L68) is there. We probably need to remove the hard-coded call to the `docker` command and replace it with a variable that's docker on cloud machines and the singularity entrypoint on HPC machines. (Useful for reference: [singularity commands](http://singularity.lbl.gov/user-guide).)

The reason I added the `set -x` line before the docker command is so the actual command the system is running is printed out. That's useful for debugging. You can try copying and pasting that command, and changing docker to singularity, to see if that (or some other) command makes things work.

(This might require manually pulling the biocontainers as well, if singularity also has a different entrypoint for a "docker pull" equivalent.)

Glad we're getting these bugs worked out! 🐛 🐞

**Comment by SichongP** [comment]
It seems that latest version of singularity supports ```singularity pull quay.io```. I'm asking our computing center to update their singularity (current installation: 2.0, latest: 2.4.5) and I'll test it once it's updated.
Although there appears to be an issue of missing /etc/resovl.conf file in quay.io biocontainer images (see [#749](https://github.com/singularityware/singularity/issues/749#issuecomment-311101264))
Not sure if it's relevant in this case.

**Comment by charlesreid1** [comment]
Okay, re singularity pull, sounds good.

Re issue 749, that looks a bit hairy. I am pretty sure @brooksph has run biocontainers using singularity on HPC, so he may have some insights. (On the other hand, it could be a problem limited to a few containers, so it may not have come up before.)

**Comment by brooksph** [comment]
Thanks for bringing this up @SichongP. I encountered some issues with singularity and quay.io a few months back but I think most of the issues have been worked out. As of a few moments ago, I'm able to successfully pull images from quayio and run the containers on the msu hpc using singularty version 2.4.

**Comment by SichongP** [comment]
I tried to run taco with singularity on my laptop but apparently singularity does not work well with WSL as of now and I'm not ready for hassle of dual booting on a windows machine yet. Please let me know if there is some magic to work around this or I'll just have to wait for our computing center to update software, which can be forever :(

**Comment by charlesreid1** [comment]
Hey Sichong, sorry to hear you've been having so much trouble getting this running on Davis HPC. But maybe we can clear this up, since I'm a little confused about what the issue actually is. 

Originally I thought it was that Snakemake was using docker instead of singularity (although I can't see the full error message you pasted in the original issue - **please follow up with full error message if possible!**). But you also mentioned something about the singularity pull command and singularity 2.0 needing to be upgraded. Is the problem that `singularity pull` is not supported in singularity 2.0, but it is supported in version 2.4? (I would be surprised if they added important commands without incrementing the major version.)

@brooksph you mentioned that these issues were worked out already - was it the same issue with running `singularity pull` with singularity 2.0, and are we all talking about the same issue?

**Comment by SichongP** [comment]
Sorry for the confusion! I'll paste full error message in next comment.

So the reason that I have to wait for them to update singularity is the current version installed on our cluster does not support running images hosted on docker://quay.io and it doesn't even have a pull command. 
I installed latest version on my laptop and it was able to pull image from docker://quay.io but it couldn't make mountspace, which after searching I think is b/c WSL does not support such functions yet.

Error message when running singularity on WSL:
```
Docker image path: quay.io/biocontainers/fastqc:0.11.7--pl5.22.0_2
Cache folder set to /home/pengsc/.singularity/docker
Creating container runtime...
tar: ./.exec: implausibly old time stamp -9223372036854775808
tar: ./.run: implausibly old time stamp -9223372036854775808
tar: ./.shell: implausibly old time stamp -9223372036854775808
tar: ./.singularity.d/actions/exec: implausibly old time stamp -9223372036854775808
tar: ./.singularity.d/actions/run: implausibly old time stamp -9223372036854775808
tar: ./.singularity.d/actions/shell: implausibly old time stamp -9223372036854775808
tar: ./.singularity.d/actions/start: implausibly old time stamp -9223372036854775808
tar: ./.singularity.d/actions/test: implausibly old time stamp -9223372036854775808
tar: ./.singularity.d/actions: implausibly old time stamp -9223372036854775808
tar: ./.singularity.d/env/01-base.sh: implausibly old time stamp -9223372036854775808
tar: ./.singularity.d/env/90-environment.sh: implausibly old time stamp -9223372036854775808
tar: ./.singularity.d/env/95-apps.sh: implausibly old time stamp -9223372036854775808
tar: ./.singularity.d/env/99-base.sh: implausibly old time stamp -9223372036854775808
tar: ./.singularity.d/env: implausibly old time stamp -9223372036854775808
tar: ./.singularity.d/libs: implausibly old time stamp -9223372036854775808
tar: ./.singularity.d/runscript: implausibly old time stamp -9223372036854775808
tar: ./.singularity.d/startscript: implausibly old time stamp -9223372036854775808
tar: ./.singularity.d: implausibly old time stamp -9223372036854775808
tar: ./.test: implausibly old time stamp -9223372036854775808
tar: ./dev: implausibly old time stamp -9223372036854775808
tar: ./environment: implausibly old time stamp -9223372036854775808
tar: ./etc/hosts: implausibly old time stamp -9223372036854775808
tar: ./etc/resolv.conf: implausibly old time stamp -9223372036854775808
tar: ./etc: implausibly old time stamp -9223372036854775808
tar: ./home: implausibly old time stamp -9223372036854775808
tar: ./proc: implausibly old time stamp -9223372036854775808
tar: ./root: implausibly old time stamp -9223372036854775808
tar: ./singularity: implausibly old time stamp -9223372036854775808
tar: ./sys: implausibly old time stamp -9223372036854775808
tar: ./tmp: implausibly old time stamp -9223372036854775808
tar: ./var/tmp: implausibly old time stamp -9223372036854775808
tar: ./var: implausibly old time stamp -9223372036854775808
tar: .: implausibly old time stamp -9223372036854775808
ERROR  : Could not make mountspaces slave: Invalid argument
ABORT  : Retval = 255
```

**Comment by SichongP** [comment]
Error message in taco run with docker:
```
--------
taco workflow details:
        snakefile: rules/read_filtering/Snakefile
        config: workflow-config/config_step2.json
        params: workflow-params/params_step2.json
        targets: ['data/SRR606249_1_reads_fastqc.zip', 'data/SRR606249_2_reads_fastqc.zip']
        bootstrap configuration: {'data_dir': 'data/', 'clean': False}
--------
Building DAG of jobs...
Using shell: /bin/bash
Provided cores: 1
Rules claiming more threads will be scaled down.
Job counts:
        count   jobs
        2       pre_trimming_quality_assessment
        2

Job 0: --- Pre-trim quality check of trimmed data with fastqc.

Building DAG of jobs...
Using shell: /bin/bash
Job counts:
        count   jobs
        1       pre_trimming_quality_assessment
        1

            set -x
            docker run                 -u "$(id -u):$(id -g)"                 -v /home/pengsc/dahak/taco-read-filtering/data/:/data/                 -it                 quay.io/biocontainers/fastqc:0.11.7--pl5.22.0_2                 fastqc                 -t 1                 /data/SRR606249_1_reads.fq.gz                 -o /data/
            set +x

++ id -u
++ id -g
+ docker run -u 20921:20921 -v /home/pengsc/dahak/taco-read-filtering/data/:/data/ -it quay.io/biocontainers/fastqc:0.11.7--pl5.22.0_2 fastqc -t 1 /data/SRR606249_1_reads.fq.gz -o /data/
/bin/bash: line 2: docker: command not found
Error in rule pre_trimming_quality_assessment:
    jobid: 0
    output: data/SRR606249_1_reads_fastqc.zip

RuleException:
CalledProcessError in line 93 of /home/pengsc/dahak/taco-read-filtering/rules/read_filtering/pre_post_assessment.rule:
Command ' set -euo pipefail;
            set -x
            docker run                 -u "$(id -u):$(id -g)"                 -v /home/pengsc/dahak/taco-read-filtering/data/:/data/                 -it                 quay.io/biocontainers/fastqc:0.11.7--pl5.22.0_2                 fastqc                 -t 1                 /data/SRR606249_1_reads.fq.gz                 -o /data/
            set +x ' returned non-zero exit status 127.
  File "/home/pengsc/dahak/taco-read-filtering/rules/read_filtering/pre_post_assessment.rule", line 93, in __rule_pre_trimming_quality_assessment
  File "/home/pengsc/miniconda3/lib/python3.6/concurrent/futures/thread.py", line 56, in run
Shutting down, this might take some time.
Exiting because a job execution failed. Look above for error message
Complete log: /home/pengsc/dahak/taco-read-filtering/.snakemake/log/2018-04-30T181410.476195.snakemake.log
Shutting down, this might take some time.
Exiting because a job execution failed. Look above for error message
Complete log: /home/pengsc/dahak/taco-read-filtering/.snakemake/log/2018-04-30T181409.981534.snakemake.log

Error: Failed to call Snakemake API
You might have requested an output file or rule that does not exist.
Try adding the data directory location as a prefix to your file names.

config['data_dir'] = data/
```

**Comment by charlesreid1** [comment]
Thanks @SichongP - that's helpful and helps clear things up. There  is actually _no documentation_ available for this version of singularity (<2.2) b/c it is so old! ([singularity docs links](http://singularity.lbl.gov/archive/))

Is this on the cabernet cluster? If we're stuck in the mud on this cluster and there's no hope of updating singularity, we may want to start thinking about whether we can find another cluster to test taco on. :-/
